File: Backend/smartbrain/src/main/java/com/smartparking/smartbrain/model/Wallet.java
package com.smartparking.smartbrain.model;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.util.Currency;
import java.util.Set;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.FieldDefaults;

@Getter
@Setter
@Entity
@Table(name = "wallets")
@ToString
@EqualsAndHashCode
@AllArgsConstructor
@NoArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Wallet {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    String walletId;

    @Column(nullable = false)
    @NotNull(message = "Balance cannot be null")
    BigDecimal balance = BigDecimal.ZERO;

    @Column(nullable = false)
    @NotNull(message = "Currency cannot be null")
    Currency currency;

    @NotNull(message = "Name of wallet cannot be null")
    String name;

    // Relationship
    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    @NotNull(message = "User cannot be null")
    User user;

    @OneToMany(mappedBy = "wallet", cascade = CascadeType.ALL, orphanRemoval = true)
    Set<Transaction> transactions;
    // Timestamp
    Timestamp createdAt;
    Timestamp updatedAt;
    public void onUpdated() {
        this.updatedAt = new Timestamp(System.currentTimeMillis());
    }

    public void onCreated() {
        this.createdAt = new Timestamp(System.currentTimeMillis());
    }

}

File: Frontend/myparkingapp/lib/data/wallet.dart
class Wallet {
  String walletId;
  double balance;
  String currency;
  String name;
  String userId;

  Wallet({
    required this.walletId,
    required this.balance,
    required this.currency,
    required this.name,
    required this.userId,
  });

  // Convert JSON to Wallet object
  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      walletId: json['walletId'] as String,
      balance: (json['balance'] as num).toDouble(),
      currency: json['currency'] as String,
      name: json['name'] as String,
      userId: json['userId'] as String,
    );
  }

  // Convert Wallet object to JSON
  Map<String, dynamic> toJson() {
    return {
      'walletId': walletId,
      'balance': balance,
      'currency': currency,
      'name': name,
      'userId':userId
    };
  }
}